### PR TITLE
Add service layer support to manage the user defined local authenticators

### PR DESCRIPTION
**Pull request by Thisara-Welmilla**
Task issue:
- https://github.com/wso2/product-is/issues/21216

With this PR following changes may introduced.
1. Add new property `authenticationType` for UserDefinedLocalAuthenticatorConfig model.
2. Improve application authentication service to support managing new local custom authenticators (Local user account identification and 2FA authentication types)
3. Add permissions and scopes related new API `/authenticators/custom/`

Related PRs:
- DAO layer implementation 
https://github.com/wso2/carbon-identity-framework/pull/6144
- Unit tests
https://github.com/wso2/carbon-identity-framework/pull/6072
- API server PR:
https://github.com/wso2/identity-api-server/pull/736

### CODE REVIEW

File: components/application-mgt/org.wso2.carbon.identity.application.common/src/main/java/org/wso2/carbon/identity/application/common/ApplicationAuthenticatorService.java [hunk 1]
```diff
@@ -51,6 +70,17 @@ public List<LocalAuthenticatorConfig> getLocalAuthenticators() {
         return this.localAuthenticators;
     }
 
+    /**
+     * This returns User defined local authenticators.
```
Comment (malithie) [inline components/application-mgt/org.wso2.carbon.identity.application.common/src/main/java/org/wso2/carbon/identity/application/common/ApplicationAuthenticatorService.java:74]: Be consistent on the terms and case using

File: components/application-mgt/org.wso2.carbon.identity.application.common/src/main/java/org/wso2/carbon/identity/application/common/ApplicationAuthenticatorService.java [hunk 2]
```diff
@@ -121,4 +187,107 @@ public void removeRequestPathAuthenticator(RequestPathAuthenticatorConfig authen
...
+     */
+    public UserDefinedLocalAuthenticatorConfig addUserDefinedLocalAuthenticator(
+            UserDefinedLocalAuthenticatorConfig authenticatorConfig, AuthenticationType type, String tenantDomain)
+            throws AuthenticatorMgtException {
+
+        LocalAuthenticatorConfig config = getLocalAuthenticatorByName(authenticatorConfig.getName(), tenantDomain);
+        if (config != null) {
+            ErrorMessages error = ErrorMessages.ERROR_AUTHENTICATOR_ALREADY_EXIST;
+            throw new AuthenticatorMgtClientException(error.getCode(), error.getMessage(),
+                    String.format(error.getDescription(), authenticatorConfig.getName()));
```
Comment (malithie) [inline components/application-mgt/org.wso2.carbon.identity.application.common/src/main/java/org/wso2/carbon/identity/application/common/ApplicationAuthenticatorService.java:208]: Shall we introduce a util to build the exception from the error ?

File: components/application-mgt/org.wso2.carbon.identity.application.common/src/main/java/org/wso2/carbon/identity/application/common/constant/AuthenticatorMgtErrorConstants.java [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.wso2.carbon.identity.application.common.constant;
+
+/**
+ * Constants for authenticator configuration management service.
+ */
+public class AuthenticatorMgtErrorConstants {
```
Comment (malithie) [inline components/application-mgt/org.wso2.carbon.identity.application.common/src/main/java/org/wso2/carbon/identity/application/common/constant/AuthenticatorMgtErrorConstants.java:24]: My suggestion is to have a util to build exceptions and within the util define all the error constants

**Comment by malithie** [comment]
Given this introduces a new API resource and scopes it should also be included in "system-api-resource.xml". So that it can be subscribed and accessed from an oauth app via console, or the respective permission to be incorporated to a user role

File: components/application-mgt/org.wso2.carbon.identity.application.common/src/main/java/org/wso2/carbon/identity/application/common/util/AuthenticatorMgtExceptionBuilder.java [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+        ERROR_AUTHENTICATOR_ALREADY_EXIST("60003", "An authenticator already exists.",
+                "As authenticator already exists for the given name: %s."),
+        ERROR_INVALID_AUTHENTICATOR_NAME("60004", "Authenticator name is invalid.",
+                "The provided authenticator name %s is not in the expected format %s."),
+        ERROR_BLANK_FIELD_VALUE("60004", "Invalid empty or blank value.",
+                "Value for %s should not be empty or blank."),
+
+        // Server errors.
+        ERROR_WHILE_ADDING_AUTHENTICATOR("65001", "Error while adding authenticator.",
+                "Error while persisting authenticator from the system."),
```
Comment (malithie) [inline components/application-mgt/org.wso2.carbon.identity.application.common/src/main/java/org/wso2/carbon/identity/application/common/util/AuthenticatorMgtExceptionBuilder.java:96]: from the system -> in system

Comment (malithie) [reply components/application-mgt/org.wso2.carbon.identity.application.common/src/main/java/org/wso2/carbon/identity/application/common/util/AuthenticatorMgtExceptionBuilder.java:96]: Fix the rest similarly

**Comment by Thisara-Welmilla** [comment]
Reopen with https://github.com/wso2/carbon-identity-framework/pull/6172

**Pull request closed without merging**